### PR TITLE
Test spack build with CI and provide instructions and tools for release

### PR DESCRIPTION
**Pull request by BenjaminRueth**
This PR aims at providing a CI pipeline for testing the spack build process using the current version of the python bindings and a `package.py` that is provided in this repository.

# Open Todo

- [x] Provide a docker image with the necessary dependencies to reduce build time

### CODE REVIEW

**Comment by BenjaminRueth** [comment]
https://github.com/precice/python-bindings/pull/65#issuecomment-727237472 provides an update for `package.py` and another `patch` that we will have to implement here as soon as #65 is merged.

**Review by BenjaminRueth — COMMENTED** [review]
Quick summary of what this PR actually does (at the current state):

### General Problem

We do not test whether the python-bindings can be built using spack. The corresponding package `py-pyprecice` is only updated for new releases. This PR adds a CI workflow in github actions for building the spack package based on a given state of the repository (`develop` or any other branch).

### Details on `build-spack.yml`

We use the image `benjaminrueth/ci-spack-precice-1804` as a basis to reduce build time. This image already provides spack and the dependencies of preCICE:

```
# Dockerfile of benjaminrueth/ci-spack-precice-1804

FROM spack/ubuntu-bionic:latest
RUN spack install boost@1.74.0 cmake@3.18.4 eigen@3.3.8 petsc@3.14.1
RUN spack env create precice
```

### What in the test?

We provide all necessary files for the spack package under `spack/var/spack/repos/builtin/packages/py-pyprecice`. The provided `package.py` and `*.patch` files are used to build the spack package based on the current state of the repository. The script `jinja-instantiate.py` allows to specify the branch to be used for `py-pyprecice@develop`.

We finally copy these files to the spack directory of `benjaminrueth/ci-spack-precice-1804` to overwrite the specification from `spack/develop` and install the package via `spack install py-pyprecice@develop`.

### Current issues

Unfortunately github actions do not pick up the dependencies provided within `benjaminrueth/ci-spack-precice-1804`, while using `act` or when running the container interactively everything looks fine. This leads to a huge build time.

**Comment by fsimonis** [comment]
I had the same issue with the precice actions. You need to use a login shell:

https://github.com/precice/precice/blob/641e5057cb5d71d39ed43ae1dbc6af597b60bf52/.github/workflows/build-and-test.yml#L20-L22

**Comment by BenjaminRueth** [comment]
### Notes for the review

@ajaust: any obvious spack flaws?
@fsimonis: what do you think about the general workflow and test? We can also consider moving some of the docker/CI related things to https://github.com/precice/ci-images, if you think this makes sense. Building the image with the dependencies, of course, takes ages, but this is something one only has to do from time to time.

File: spack/ci-spack-pyprecice-deps-1804.dockerfile
Comment (ajaust) [inline spack/ci-spack-pyprecice-deps-1804.dockerfile:10]: Why are you installing pkgconfig in line 6 and 7?

I think the cleanest way of compiling/adding the dependencies to the environment and concretizing them trying to avoid duplication would be:
```
RUN spack env activate precice \ 
    && spack add boost@1.74.0 cmake@3.18.4 eigen@3.3.8 petsc@3.14.1 pkgconfig precice@develop py-setuptools py-wheel py-cython py-numpy py-mpi4py \
    && spack concretize -f \
    && spack install
```

You might specify the environment in a `spack.yaml` file in `${SPACK_ROOT}/var/spack/environments/precice` (where `precice` is the name of the environment). It might help to add the line `concretization: together` in case dependencies are duplicated.

The `spack.yaml` might look like this:
```
# This is a Spack Environment file.
#
# It describes a set of packages to be installed, along with
# configuration settings.
spack:
   specs:
  - boost@1.74.0 
  - cmake@3.18.4 
  - eigen@3.3.8 
  - petsc@3.14.1
  - pkgconfig
  - precice@develop
  - py-setuptools 
  - py-wheel 
  - py-cython 
  - py-numpy
  - py-mpi4py
  view: true
  concretization: together
```

If it works for now, I guess it should be fine. I am not sure if @fsimonis  has an opinion on how to set environments up properly.

File: spack/var/spack/repos/builtin/packages/py-pyprecice/package.py
Comment (ajaust) [inline spack/var/spack/repos/builtin/packages/py-pyprecice/package.py:64]: Don't we need this for older versions of the bindings? 

If yes, I guess we should try to fix it properly. If we do not need that for older versions, it should be removed.

**Review by ajaust — COMMENTED** [review]
I have added some comments, but it looks quite good to me.

File: spack/README.md
Comment (BenjaminRueth) [reply spack/README.md]: Done 4299c8c.

File: spack/README.md
Comment (BenjaminRueth) [reply spack/README.md:17]: Done 4299c8c.

File: spack/ci-spack-pyprecice-deps-1804.dockerfile
Comment (BenjaminRueth) [reply spack/ci-spack-pyprecice-deps-1804.dockerfile]: I changed the dockerfile in 2e22d8c. I will rebuild and test the new file in a moment. But let's keep this threat open until @fsimonis has finished the review. I'm not very experienced with spack and therefore another opinion on the question on "how do I setup an environment?" probably does not harm.

Comment (BenjaminRueth) [reply spack/ci-spack-pyprecice-deps-1804.dockerfile]: Note on `pkgconfig`: Currently preCICE needs `pkgconfig` for finding PETSc. As far as I understand the issue this dependency is currently missing in the package definition of `precice`. Therefore, I'm installing it manually.

File: spack/var/spack/repos/builtin/packages/py-pyprecice/package.py
Comment (BenjaminRueth) [reply spack/var/spack/repos/builtin/packages/py-pyprecice/package.py:64]: Totally forgot about this. 100% true. I will try to find out how to implement this in `package.py`. I think we will need it for supporting older version.

File: spack/ci-spack-pyprecice-deps-1804.dockerfile
Comment (ajaust) [reply spack/ci-spack-pyprecice-deps-1804.dockerfile]: I guess I missed that this is Docker and every `RUN` is executed in a new container (or so...). So I assume the sole
```
RUN spack install pkgconfig 
```
will install the package in the root environment and 
```
spack env activate precice && spack concretize -f && spack install pkgconfig && spack install precice@develop
```
adds it to the `precice` environment. Regarding the resulting software this should not make any difference. Software that is installed in an environment ends up in the root environment as well.

File: spack/var/spack/repos/builtin/packages/py-pyprecice/package.py
Comment (BenjaminRueth) [reply spack/var/spack/repos/builtin/packages/py-pyprecice/package.py]: d981f41 implements support for older versions. However, testing this is not so easy right now, since I cannot easily check for the version (#50).

File: spack/var/spack/repos/builtin/packages/py-pyprecice/remove-packaging-and-pip.patch
Comment (BenjaminRueth) [inline spack/var/spack/repos/builtin/packages/py-pyprecice/remove-packaging-and-pip.patch:21]: We will have to review this patch, when #63 and #70 are merged.

Comment (BenjaminRueth) [reply spack/var/spack/repos/builtin/packages/py-pyprecice/remove-packaging-and-pip.patch]: Updated the patch according to #70 in 4e88ef6.

Comment (BenjaminRueth) [reply spack/var/spack/repos/builtin/packages/py-pyprecice/remove-packaging-and-pip.patch]: Removed the patch, since it is not needed anymore. 50a0196

File: spack/var/spack/repos/builtin/packages/py-pyprecice/package.py
Comment (ajaust) [inline spack/var/spack/repos/builtin/packages/py-pyprecice/package.py:61]: Does this work nownewer versions of the adapter or does it skip the installation step completely then?

File: spack/ci-spack-pyprecice-deps-1804.dockerfile
Comment (ajaust) [inline spack/ci-spack-pyprecice-deps-1804.dockerfile:8]: Is this acceptable with regards to compile time or do we need to create a Docker image that already contains all the dependencies?

**Review by ajaust — COMMENTED** [review]
I have added my feedback, but that is only minor things.

File: spack/ci-spack-pyprecice-deps-1804.dockerfile
Comment (BenjaminRueth) [reply spack/ci-spack-pyprecice-deps-1804.dockerfile:8]: The image is compiled locally (once) an provided on dockerhub (https://hub.docker.com/r/benjaminrueth/ci-spack-pyprecice-deps-1804). This Dockerfile is only provided in case we want to rebuild the image.

File: spack/var/spack/repos/builtin/packages/py-pyprecice/package.py
Comment (BenjaminRueth) [reply spack/var/spack/repos/builtin/packages/py-pyprecice/package.py:61]: The current test runs on `develop > 2.1.1.1` and works. I ran an additional tests and to me it looks like `install` does what it is supposed to do:

Running `python3 -c "import precice; print(precice.__version__)"` not from project root, but from an independent folder works. So the package is found correctly (through the python installation), which means that the installation was executed correctly.

If I run `python3 -c "import precice; print(precice)"`, I also get a reasonable path: `<module 'precice' from '/opt/spack/var/spack/environments/precice/.spack-env/view/lib/python3.8/site-packages/precice/__init__.py'>`

I think this looks good from my perspective. We should still explicitly ask for a double check on this part, when we submit the new `package.py` to Spack, since our reviewer will probably know the answer. I myself are not 100% sure.

Comment (ajaust) [reply spack/var/spack/repos/builtin/packages/py-pyprecice/package.py:61]: That sounds good!

File: spack/ci-spack-pyprecice-deps-1804.dockerfile
Comment (ajaust) [reply spack/ci-spack-pyprecice-deps-1804.dockerfile:8]: That sounds good, too. I was not aware of the actual workflow.

**Review by fsimonis — APPROVED** [review]

**Pull request merged**
